### DISCUSSION TITLE
[v5] How run migration on `node ace start`

### DISCUSSION THREAD

**Discussion by satheler** [adonisjs/core#2429] [Help] [answered]
Is there any way to run the migration when starting the server, without having to run `node ace migration:run`? Maybe something on `server.ts`?

**Comment by thetutlage** [comment]
Can you please share the use case you have. Since running migrations on server start is not something I recommend at all

**Reply by satheler** [reply]
I don't want run migration on my CI/CD pipeline. 
I prefer run the migration on server starts, even though he knows that he will not run any migration.

This approach that I am proposing is something similar to the `sequelize sync`.

**Reply by satheler** [reply]
My quick solution was to use the execa library and run the command `node ace migration: run --force` (NODE_ENV is production).

However, I would like something more performative.

**Reply by McSneaky** [reply]
It has many-many downsides, but you can just add it to startup command
`node ace migration:run --force && node server.js`

**Reply by satheler** [reply]
But in some cases I don't have access to environment.

**Comment by ivan-prats** [comment, marked as answer]
Hello! I am using Render to deploy my Adonis 5 app (PaaS similar to Heroku). 

I would assume that usually the best way to run migrations in a more traditional server is to:
1. Notify my customers of a minor maintenance downtime at XXX (that should be maximum of a couple minutes supposing that everything goes smoothly)
2. Stop the `server.ts` thread at said time and run the migrations (previously tested and with proper down() method in case we want to revert them), 
3. Check everything is alright with the production database
4. Restart the server, possibly with new code that will use the new migrations that have been run

The issue with Render is that I don't think there is any way to stop the `servert.ts` without killing completely the server. I can "delete" or "suspend" the service, but that does not allow me to run any commands while it's suspended (the server has been brought down completely), making it impossible to run the migrations while it is suspended.

So in this case I have the following options left to run migrations in production:
1. Like @satheler said: run automatically the migrations with the server start. Deploying my new code alongside with the migration 
2. Running the migrations remotely from my development machine while targetting the production database. After running them and checking everything is alright: deploying the new code to the server.

I share your (@thetutlage , @McSneaky ) opinion that the first option has many many downsides and risks, so I end up going with the second one. But that option is also not great, because it forces me to write the migrations making sure that they are both compatible with the previous and new code I will be deploying. 

Is there any other option I haven't thought about? How do other teams running in Heroku run their migrations on production?

**Reply by McSneaky** [reply]
You can use CI for deployment (Github Actions / Gitlab CI / Jenkins etc) to do the deployment and then post-deployment script is running migrations. That's how I usually do it

**Reply by ivan-prats** [reply]
I suspect that in order to do that, there needs to be some kind of programatic access to the shell of the server, in order for a tool (or myself) to run the command. 

But I don't think the provider I'm currently using, [render](https://render.com/), has programmatic access to the shell of the application. You need to access the shell manually from their dashboard.

What providers do you usually use to deploy AdonisJS applications on production? I don't mind paying a bit extra in order to minimize the devops (hence why I like Render / Heroku type of solutions)

**Reply by McSneaky** [reply]
Aah, didn't know you don't have shell access

I'm real oldschool guy. I've been running my own VPSes on Contabo and Vultr mostly and then using Gitlab CI to automate it all

I've heard a lot of good about Cleavr tho. Tested out their service too but imo their pricing was too much. They constantly offer Discounts in AdonisJS Discord and their price is negioatable as much as I've understood. Might even be free for open-source stuff, might be worth to get into contact with them

Also if you want to do some more hacking there's Dokku and Rancher. Have had plans to test out Dokku for ages. Using Rancher quite daily to do some automations for me, coz managing K8 yourself is quite a pain

**Reply by ivan-prats** [reply]
Thanks for your help @McSneaky ! Will take a look at the options

**Reply by satheler** [reply]
> You can use CI for deployment (Github Actions / Gitlab CI / Jenkins etc) to do the deployment and then post-deployment script is running migrations. That's how I usually do it

Thanks @McSneaky! This can be a great option!
